src/deluge/gui/views/midi_session_view.cpp

```diff
@@ -0,0 +1,949 @@
+			//if yes, display parameter name and learned status
+			potentialShortcutPadAction(xDisplay, yDisplay);
+		}
+		//let go of pad
+		else {
+			renderViewDisplay();
+			initPadPress(lastPadPress);
+		}
+	}
+	return ActionResult::DEALT_WITH;
+}
+
+//check if pad press corresponds to shortcut press
+void MidiSessionView::potentialShortcutPadAction(int32_t xDisplay, int32_t yDisplay) {
+	Param::Kind paramKind = Param::Kind::NONE;
+	int32_t paramID = kNoSelection;
+
+	if (patchedParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+		paramKind = Param::Kind::PATCHED;
+		paramID = patchedParamShortcuts[xDisplay][yDisplay];
+	}
+	else if (unpatchedParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+		paramKind = Param::Kind::UNPATCHED_SOUND;
+		paramID = unpatchedParamShortcuts[xDisplay][yDisplay];
+	}
+	else if (globalEffectableParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+		paramKind = Param::Kind::UNPATCHED_GLOBAL;
+		paramID = globalEffectableParamShortcuts[xDisplay][yDisplay];
+	}
+	if (paramKind != Param::Kind::NONE) {
+		//if pressing a param shortcut while holding learn, unlearn midi CC from a specific param
+		if (Buttons::isButtonPressed(deluge::hid::button::LEARN)) {
+			paramToCC[xDisplay][yDisplay] = kNoSelection;
+			previousKnobPos[xDisplay][yDisplay] = kNoSelection;
+			updateMappingChangeStatus();
+			uiNeedsRendering(this);
+		}
+		renderParamDisplay(paramKind, paramID, paramToCC[xDisplay][yDisplay]);
+		lastPadPress.isActive = true;
+		lastPadPress.xDisplay = xDisplay;
+		lastPadPress.yDisplay = yDisplay;
+		lastPadPress.paramKind = paramKind;
+		lastPadPress.paramID = paramID;
+	}
+}
+
+void MidiSessionView::learnCC(int32_t channel, int32_t ccNumber) {
+	if (channel == midiEngine.midiFollowChannelParam) {
+		if (lastPadPress.isActive) {
+			if (paramToCC[lastPadPress.xDisplay][lastPadPress.yDisplay] != ccNumber) {
+				//init knobPos for current param
+				previousKnobPos[lastPadPress.xDisplay][lastPadPress.yDisplay] = kNoSelection;
+
+				//look to see if this cc was mapped elsewhere and we have a previousKnobPosition saved
+				for (int32_t xDisplay = 0; xDisplay < kDisplayWidth; xDisplay++) {
+					for (int32_t yDisplay = 0; yDisplay < kDisplayHeight; yDisplay++) {
+						if ((paramToCC[xDisplay][yDisplay] == ccNumber)
+						    && (previousKnobPos[xDisplay][yDisplay] != kNoSelection)) {
+							previousKnobPos[lastPadPress.xDisplay][lastPadPress.yDisplay] =
+							    previousKnobPos[xDisplay][yDisplay];
+						}
+					}
+				}
+
+				//assign cc to current param selected
+				paramToCC[lastPadPress.xDisplay][lastPadPress.yDisplay] = ccNumber;
+
+				renderParamDisplay(lastPadPress.paramKind, lastPadPress.paramID, ccNumber);
+
+				currentCC = kNoSelection;
+				updateMappingChangeStatus();
+			}
+		}
+		else {
+			currentCC = ccNumber;
+		}
+		uiNeedsRendering(this);
+	}
+	else if (lastPadPress.isActive) {
+		cantLearn(channel);
+	}
+}
+
+void MidiSessionView::cantLearn(int32_t channel) {
```

Comment:
add documentation - error message displayed when you try to learn a midi cc to a param on a channel different from the midi follow param channel